### ISSUE TITLE
Extract Cabal-Helper specific code into its own library

### ISSUE THREAD

**Issue by fendor** [haskell/haskell-ide-engine#1631]
Talking with @jneira we realized this makes it easier to share code between the projects!
Opinions? Code would be entirely from `Cradle.hs` in `hie-plugin-api`. The entire logic of discovering cradles would be moved into its own library.

**Comment by jneira** [comment]
I've just discovered that [`Haskell.IDe.Engine.Cradle`](https://github.com/haskell/haskell-ide-engine/blob/master/hie-plugin-api/Haskell/Ide/Engine/Cradle.hs) only imports `Haskell.Ide.Engine.Logger`!
Almost it is asking to live in its own home. 😄

**Comment by jneira** [comment]
In case we decided to creae the library, what about `implicit-hie-bios` (or `hie-bios-implicit` or something alike), to not bind it to `cabal-helper`.
Maybe some day both build tools stack and cabal gives us enough info to use them directly?

**Comment by ndmitchell** [comment]
I really like the abstraction boundary that hie-bios figures out what the flags for the session should be. If it became a plugin API that I could then configure I'd be much less happy, as it's something I'm super grateful to not have to care about. Whether it incorporates or includes cabal-helper like functionality under the hood is a detail, so if this is transparent, I'm fine with it.

**Comment by fendor** [comment]
I would not have considered it a plugin API, but I think I see what you mean.
The intention would have been to extract code from [`Haskell.IDe.Engine.Cradle`](https://github.com/haskell/haskell-ide-engine/blob/master/hie-plugin-api/Haskell/Ide/Engine/Cradle.hs) since only roughly six functions are used elsewhere in haskell-ide-engine, none of them specific to cabal-helper.
I think one can argue that this micro-package would be nothing else but moved code for now. 
Do you dislike the idea to have an independent package for cabal-helper to hie-bios cradle conversion? Does that already count as a plugin API?

**Comment by ndmitchell** [comment]
Is the cabal-helper to hie-bios cradle something ghcide has to mix in? I don't like that idea. Or something that hie-bios always links in? If so, I don't see the advantage of separating the package.

**Comment by fendor** [comment]
No, currently only hie uses it and hls currently copied the code to https://github.com/haskell/haskell-language-server/blob/master/src/Ide/Cradle.hs
It is an alternative implementation to using the `implicitCradle` function. It is optional and would not affect ghcide at all, unless ghcide decides to use cabal-helper as well.

**Comment by ndmitchell** [comment]
I think hls and ghcide should converge on this as much as possible.

**Comment by fendor** [comment]
Absolutely agreed. I think, somewhere it was mentioned that you agreed in bristol to use cabal-helper? Should there be an attempt to use cabal-helper in ghcide?

**Comment by jneira** [comment]
@ndmitchell just in case it would be solvable, what are your concerns about link statically cabal-helper in ghcide? is it that it depends on the Cabal library or another problematic lib? or the runtime compilation it triggers?

If hls uses cabal helper directly it will need to pass the cradle generated by c-h to ghcide (the component that actually is using it directly). Not sure if ithat is already possible in ghcide or it should be changed to accept "external" cradles.

Obviously if we add support in ghcide both would have the implicit cradle available. Imo it would be the desirable option over the former one.

Other options i can think of:
* discover another, more lightweight way to generate a reliable implicit cradle
* make hie-bios use cabal-helper at runtime (i.e. calling c-h as a executable), creating another cradle type (implicit?). Not sure it is possible

**Comment by ndmitchell** [comment]
I've heard cabal-helper takes a lot of work to support. At the moment, if there's an issue with the GHC flags, I go to hie-bios because the contract is hie-bios provides the flags to start a GHC session. I like the abstract interface "give me the flags" and the support model "ping Matthew et al". It's as much lines of communication and separation of responsibilities as it is code. If you guys want to shove cabal-helper behind the same interface and support model, provided there is no GPL impact, I'm happy (and moreover, consider it none of my business).

**Comment by fendor** [comment]
That is exactly how we integrated cabal-helper. It is hidden within a `Cradle` and you just query it, like any other cradle. Granted, the mechanism is not as light-weight as the implicit cradle discovery in hie-bios, but we have no extra logic in hie to support cabal-helper.
That's why usage of cabal-helper is completely isolated in this one module.

**Comment by alanz** [comment]
FYI https://github.com/haskell/haskell-language-server/blob/master/src/Ide/Cradle.hs

**Comment by alanz** [comment]
Which should end up wherever we agree

**Comment by jneira** [comment]
Well i think the consesus had been using the cabal-helper cradle in haskell-language-server, cause cradle setup is in the main exe module owned by h-l-s: https://github.com/haskell/haskell-language-server/issues/38#issuecomment-582789371
@fendor could we close this or you have some concerns about?

**Comment by fendor** [comment]
I still think outsourcing it into different package does no harm and makes it easier to support both hls and hie.
As I read it, there was no argument against it?

**Comment by jneira** [comment]
@fendor no other than the usual overhead of maintain a separate lib: github repo, maybe upload to hackage/stackage, etc and the fact hie will be deprecated at some point.
Otoh it could be used in another projects.
If you are fine setting up the new package i am too :wink:

**Comment by jneira** [comment]
Other option could be integrate the cabal-helper cradle in cabal-helper itself, if @DanielG agree it is a good idea add hie-bios as dependency.

**Comment by fendor** [comment]
I think, for now I would prefer it as a separate library. I feel like it does not really belong into cabal-helper. However, lets wait for @DanielG opinion :)
